Fix cart rules restrictions

Sollsy opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.2.x
| Description?      | Fixes issue with cart rule restrictions. After changes works similarly as specific price rules where product must match all rules from rule group instead of just one but allows you to make more specific cart rule restrictions. tested on 1.7.8.x / 8.2.x
| Type?             | bug fix
| Category?         | FO, BO 
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | create similar cart rules as on image below
| UI Tests          |  
| Fixed issue or discussion?     | Fixes #15525, #26528 
| Related PRs       |  
| Sponsor company   |  

1.7.8.x
![cartrule](https://github.com/user-attachments/assets/1889a20d-bd78-4400-b6d9-2936244f4e1d)

8.2.x

![cartrule8 2](https://github.com/user-attachments/assets/f9d9bfb1-05f2-4370-8670-ca1bcbea36bb)

This is my first time using github. I apologize if i put this inside wrong branch.

This is know issue with cart rules since forever. With these changes you can make more restrictive cart rules.

Sollsy committed:
Fix restrictive group rules CartRule.php

Fixes product group rules inside cart rules. After changes works same as specific price rules.

Sollsy committed:
Update product_rule_group.tpl

Sollsy committed:
Update conditions.tpl

ps-jarvis commented:
<!-- PR_WELCOME -->
Hello @Sollsy!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

ps-jarvis commented:
<!-- PR_WORDING -->
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Shop.Notifications.Error`
    - [ ] `Unknown type of product restriction`

(Note: this is an automated message, but answering it will reach a real human)

Sollsy committed:
Update CartRule.php

Sollsy committed:
Update CartRule.php

Hlavtox commented on classes/CartRule.php:
```diff
@@ -1066,56 +1066,40 @@ public function checkValidity(Context $context, $alreadyInCart = false, $display
+                        $eligible_products_list[] = (int) $product['id_product'] . '-' . (int) $product['id_product_attribute'];
+                    }
+                }
+                if (!count($eligible_products_list)) {
```
Why do you move it here? It will have to get products for each product group for no reason, not cached possibly. There is a reason it was outside the loop.

Hlavtox reviewed (changes requested):
Please do not remove code comments we added, now again its hard to understand what the code is doing. :-)

+ cs fix is needed

Sollsy commented:
> Please do not remove code comments we added, now again its hard to understand what the code is doing. :-)
> 
> * cs fix is needed

Sorry it was mean to 8.2.x branch not the develop.

Sollsy committed:
Update CartRule.php

Sollsy committed:
Update CartRule.php

Fix filtering so it doesnt remove eligible_products before all rules were checked.

Sollsy committed:
Update CartRule.php

display error in cart when products doesnt match any restrictions
